compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotter.kt

```diff
@@ -108,114 +154,67 @@ object ClassSnapshotter {
-                    CLASS -> isPrivate || isLocal || isAnonymous || isSynthetic
-                    SYNTHETIC_CLASS -> true
-                    // We're not sure about the other kinds of Kotlin classes, so we assume it's accessible (see this method's kdoc)
-                    else -> false
-                }
-            } else {
-                isPrivate || isLocal || isAnonymous || isSynthetic
-            }
-        }
-
-        val classIsInaccessible: MutableMap<BasicClassInfo, Boolean> = HashMap(classesInfo.size)
-        val classIdToClassInfo: Map<ClassId, BasicClassInfo> = classesInfo.associateBy { it.classId }
+    private val unixStyleRelativePaths: List<String>
 
-        fun BasicClassInfo.isTransitivelyInaccessible(): Boolean {
-            classIsInaccessible[this]?.let { return it }
+    private val closer = Closer.create()
 
-            val inaccessible = if (isInaccessible()) {
-                true
-            } else classId.outerClassId?.let { outerClassId ->
-                classIdToClassInfo[outerClassId]?.isTransitivelyInaccessible()
-                // If we can't find the outer class from the given list of classes (which could happen with faulty jars), we assume that
-                // the class is accessible (see this method's kdoc).
-                    ?: false
-            } ?: false
-
-            return inaccessible.also {
-                classIsInaccessible[this] = inaccessible
-            }
+    init {
+        root = if (directoryOrJar.isDirectory) {
+            directoryOrJar.toPath()
+        } else {
+            check(directoryOrJar.isFile && directoryOrJar.path.endsWith(".jar", ignoreCase = true))
+            FileSystems.newFileSystem(directoryOrJar.toPath(), null as ClassLoader?)
+                .also { closer.register(it) }
+                .rootDirectories.single()
         }
-
-        return classesInfo.filterTo(mutableSetOf()) { it.isTransitivelyInaccessible() }
+        unixStyleRelativePaths = Files.walk(root).use {
+            // Use `Set` to de-duplicate entries
+            it.collect(Collectors.toSet())
+        }.map {
+            val relativePath = it.root.relativize(it).toString()
+            File(relativePath).invariantSeparatorsPath
+        }.sorted()
     }
-}
-
-/** Utility to read the contents of a directory or jar. */
-private object DirectoryOrJarContentsReader {
 
     /**
-     * Returns a map from Unix-style relative paths of entries to their contents. The paths are relative to the given container (directory
-     * or jar).
-     *
-     * The map entries need to satisfy the given filter.
+     * Returns the Unix-style relative paths of all entries under the containing directory or jar which satisfy the given [filter].
      *
-     * The map entries are sorted based on their Unix-style relative paths (to ensure deterministic results across filesystems).
+     * The entries are in Unix style and are sorted to ensure deterministic results across platforms.
```

Comment:
`getUnixStyleRelativePaths` does not seem to sort them?

Reply:
It used a private variable `unixStyleRelativePaths` which was sorted (line 191 above), but I've removed it now so it's clearer.